pkg/indexerpoc/task.go

```diff
@@ -0,0 +1,449 @@
+
+		// Save the updated state
+		if err := t.db.SaveState(ctx, t.state); err != nil {
+			return fmt.Errorf("saving state after reorg: %w", err)
+		}
+
+		// Commit the transaction with our reorg handling
+		if err := tx.Commit(); err != nil {
+			return fmt.Errorf("committing transaction after reorg: %w", err)
+		}
+
+		// Return ErrReorg to signal we need to process again
+		// The next cycle will start from block X-1 and will detect if it was also reorged
+		return ErrReorg
+	}
+
+	// Process logs
+	if len(logs) > 0 {
+		if err := t.contract.ProcessLogs(ctx, logs); err != nil {
+			return fmt.Errorf("processing logs on %s: %w", t.state.NetworkName, err)
+		}
+	}
+
+	// Get the hash of the last block we processed
+	lastBlockHash, err := t.src.GetBlockHash(ctx, endBlock)
+	if err != nil {
+		return fmt.Errorf("getting hash for last block on %s: %w", t.state.NetworkName, err)
+	}
+
+	// Update state with the latest processed block
+	t.state.BlockNumber = endBlock
+	t.state.BlockHash = lastBlockHash
+
+	// Save updated state
+	if err := t.db.SaveState(ctx, t.state); err != nil {
+		return fmt.Errorf("saving state: %w", err)
+	}
+
+	// Commit transaction
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("committing transaction: %w", err)
+	}
+
+	slog.InfoContext(ctx, "indexed blocks",
+		"contract", t.contract.GetName(),
+		"network", t.state.NetworkName,
+		"start", startBlock,
+		"end", endBlock,
+		"logs", len(logs),
+	)
+
+	return nil
+}
+
+// nextBatch retrieves a batch of blocks and logs while checking for reorgs.
+func (t *task) nextBatch(
+	ctx context.Context,
+	startBlock, endBlock uint64,
+) ([]types.Log, bool, error) {
+	// Special case - if we have no previous state, no reorg check needed
+	if t.state.BlockNumber == 0 || t.state.BlockHash == (common.Hash{}) {
+		logs, err := t.src.GetLogs(ctx, startBlock, endBlock, t.filter)
+		return logs, false, err
+	}
+
+	// We need to check for reorgs first by verifying the parent hash of the first block
+	firstBlock, err := t.src.GetBlockByNumber(ctx, startBlock)
+	if err != nil {
+		return nil, false, fmt.Errorf("getting first block %d for reorg check on %s: %w",
+			startBlock, t.state.NetworkName, err)
+	}
+
+	// Check if the parent hash matches our stored hash.
+	if firstBlock.ParentHash() != t.state.BlockHash {
+		// Reorg detected!
+		return nil, true, nil
+	}
+
+	// No reorg, retrieve logs
+	// For small ranges or if concurrency is 1, use a simple approach
+	if t.concurrency <= 1 || endBlock-startBlock < 10 {
+		logs, err := t.src.GetLogs(ctx, startBlock, endBlock, t.filter)
+		if err != nil {
+			return nil, false, fmt.Errorf("getting logs on %s: %w", t.state.NetworkName, err)
+		}
+		return logs, false, nil
+	}
+
+	// For larger ranges, use concurrent processing
+	return t.getLogs(ctx, startBlock, endBlock)
+}
+
+// getLogs retrieves logs in parallel for better performance with large batches.
+// This is likely going to be used only when backfilling.
+func (t *task) getLogs(
```

Comment:
I know that the indexer sometimes takes a bit, but the multi-threading code is more complicated. The sort at the end is somewhat expensive. Do we know whether it is really worth the maintenance?